e2e-tests/resources/authd/browser_window.py

```diff
@@ -0,0 +1,193 @@
+
+        self.load_state = WebKit.LoadEvent.STARTED
+
+        def on_load_changed(_, load_event):
+            self.load_state = load_event
+
+        self.web_view.connect("load-changed", on_load_changed)
+
+        self._overlay = Gtk.Overlay()
+        self._overlay.add(self.web_view)
+
+        self.add(self._overlay)
+
+    def wait_for_page_loaded(self):
+        if self.load_state == WebKit.LoadEvent.FINISHED:
+            return
+
+        loop = GLib.MainLoop()
+
+        def on_load_changed(_, load_event):
+            if load_event != WebKit.LoadEvent.FINISHED:
+                return
+
+            loop.quit()
+
+        signal_id = self.web_view.connect("load-changed", on_load_changed)
+        loop.run()
+        self.web_view.disconnect(signal_id)
+
+    def wait_for_stable_page(self):
+        self.wait_for_page_loaded()
+
+        overlay = Gtk.Button()
+        overlay.set_opacity(0)
+        self._overlay.add_overlay(overlay)
+        overlay.show()
+        overlay.grab_focus()
+
+        loop = GLib.MainLoop()
+
+        def on_timeout():
+            loop.quit()
+            return False
+
+        draw_timeout = 500
+        timeout = GLib.timeout_add(draw_timeout, on_timeout)
+
+        def on_draw_event(_, de):
+            nonlocal timeout
+
+            GLib.source_remove(timeout)
+            timeout = GLib.timeout_add(draw_timeout, on_timeout)
+            return False
+
+        signal_id = self.web_view.connect("draw", on_draw_event)
+        loop.run()
+        self.web_view.disconnect(signal_id)
+
+        overlay.destroy()
+        self.web_view.grab_focus()
+
+    def wait_for_text_visible(self, text, timeout_ms=5000,
+                              poll_interval_ms=100):
+        """Wait until `text` is present in the page's visible text or raise TimeoutError."""
+        loop = GLib.MainLoop()
+        poll_id = None
+        timeout_id = None
+        found = False
+
+        def on_js_finished(web_view, result, user_data):
+            nonlocal poll_id, timeout_id, found
+            try:
+                res = web_view.run_javascript_finish(result)
+                js_value = res.get_js_value()
+                found = bool(js_value.to_boolean())
+            except Exception:
+                found = False
+
+            if found:
+                if timeout_id:
+                    GLib.source_remove(timeout_id)
+                if poll_id:
+                    GLib.source_remove(poll_id)
+                loop.quit()
+
+        def poll_fn():
+            # Use json.dumps to safely escape the text into a JS string literal
+            js = (
+                     "Boolean(document.body && document.body.innerText && "
+                     "document.body.innerText.indexOf(%s) !== -1)"
+                 ) % json.dumps(text)
+            self.web_view.run_javascript(js, None, on_js_finished, None)
+            return True  # keep polling until callback quits the loop
```

Comment:
Instead of doing this here, I'd return `False` here and instead restart the polling (calling `poll_id = GLib.timeout_add(poll_interval_ms, poll_fn)` after the finish function has returned, so that there's no possibility that we're polling twice while the same async function has not returned yet.